### ISSUE TITLE
Edge in P2P mode registers with its own IP causing packet flood

### ISSUE THREAD

**Issue by struanb** [ntop/n2n#766]
Running edge v2.9.0 without `-S1` can result in the edge IP being registered with the supernode, and inter-edge packets being sent over the edge network itself, resulting in a packet flood that saturates the N2N network.

i.e. On charlie:
`supernode -v -F FederationSecret -f -p 11113 -a 192.168.6.0-192.168.6.0/24`
`edge -v -f -c Test -k Secret -d edge1 -l charlie:11113 -r -p 11114 -a 192.168.6.5`

On domino:
`edge -v -f -c Test -k Secret -d edge1 -l charlie:11113 -r -p 11114 -n 192.168.2.0/24:192.168.6.5 -a 192.168.6.1`

```
edge -v -f -c Test -k Secret -d edge1 -l charlie:11113 -r -p 11114 -a 192.168.6.5
<snip>
20/Aug/2021 21:08:58 [edge_utils.c:2464] [P2P] Rx REGISTER from 192.168.6.183:11114
20/Aug/2021 21:08:58 [edge_utils.c:1362] send REGISTER_ACK 192.168.6.183:11114
20/Aug/2021 21:08:58 [edge_utils.c:1317] Send REGISTER to 192.168.6.183:11114
20/Aug/2021 21:08:58 [edge_utils.c:1317] Send REGISTER to 224.0.0.68:1968
20/Aug/2021 21:08:58 [edge_utils.c:1317] Send REGISTER to 192.168.2.3:11113
<snip>
20/Aug/2021 21:09:08 [edge_utils.c:1317] Send REGISTER to 192.168.6.183:11114
20/Aug/2021 21:09:08 [edge_utils.c:1317] Send REGISTER to 224.0.0.68:1968
<snip>
20/Aug/2021 21:09:08 [edge_utils.c:1317] Send REGISTER to 192.168.6.183:11114
20/Aug/2021 21:09:08 [edge_utils.c:1317] Send REGISTER to 224.0.0.68:1968
20/Aug/2021 21:09:08 [edge_utils.c:1317] Send REGISTER to 192.168.6.183:11114
20/Aug/2021 21:09:08 [edge_utils.c:1317] Send REGISTER to 224.0.0.68:1968
20/Aug/2021 21:09:08 [edge_utils.c:1317] Send REGISTER to 192.168.6.183:11114
20/Aug/2021 21:09:08 [edge_utils.c:1317] Send REGISTER to 224.0.0.68:1968
20/Aug/2021 21:09:08 [edge_utils.c:1317] Send REGISTER to 192.168.6.183:11114
20/Aug/2021 21:09:08 [edge_utils.c:1317] Send REGISTER to 224.0.0.68:1968
20/Aug/2021 21:09:08 [edge_utils.c:1317] Send REGISTER to 192.168.6.183:11114
20/Aug/2021 21:09:08 [edge_utils.c:1317] Send REGISTER to 224.0.0.68:1968
20/Aug/2021 21:09:08 [edge_utils.c:1317] Send REGISTER to 192.168.6.183:11114

**Comment by Logan007** [comment]
Thank you for bringing it up as separate issue helping us to better keep track of it.

I have had a first look into the code when we discussed this earlier at #562 and I still see no explanation for this extremely unintended behavior. I will have as deeper look. Strange this is...

As mentioned over at #753, `-b` will solve it but restrict your edge's binding to one IP address only. It might be sufficient for most cases.

Maybe we can enhance `-b` to multi-`-b`, i.e. allowing specify several IP addresses to bind to, excluding the TAP's one of course. But, it should not be the user needing to worry about it...

**Comment by Logan007** [comment]
In addition to or instead of binding edges to one IP address, you can try the freshly fixed `-DSKIP_MULTICAST_PEERS_DISCOVERY` [macro / compile time option](https://github.com/ntop/n2n/blob/dev/doc/Building.md#disable-multicast-local-peer-detection) to disable multicast peer detection. You probably do not need the multicast detection anyway if using `-f` from the `prefSock` branch which I have just rebased to also include the fixed macro handling.

I assume this to work better because the multicast usually binds to every interface which can't happen if completely disabled. I am sorry that for now we do not have a cli option but only the more complicated compile time macro to disable it.

**Comment by fengdaolong** [comment]
This form is not elegant enough.

**Comment by Logan007** [comment]
> This form is not elegant enough.

Well, it's what we have had up to now, just fixed the macro use. You probably prefer a command line option?

**Comment by fengdaolong** [comment]
`SKIP_MULTICAST_PEERS_DISCOVERY` was originally added temporarily to shield the windows multicast problem, and it should be removed now. I think for this problem, it may be better to modify the `-E` option.

**Comment by Logan007** [comment]
This definitely is not the final state, just a real quick fix to get an edge without multicast. peer detetction. 

`-E` and the SKIP_MULTICAST_PEERS_DISCOVERY macro aim at different levels (overlay vs. underlying, real network) so I would no want to mix that. But we could add a different command line option in the next step if these tests prove useful. Looking for some matching letter... :wink:

**Comment by fengdaolong** [comment]
There are too many options in n2n, and it seems too complicated, add a great cognitive burden to users. Would it be better if similar and unimportant functions are merged into the same option?

For this problem, if the `-E` option is not added, do not enable the multicast socket `eee->udp_multicast_sock`, and then turn off all multicast-related functions. If the `-E` option is added, and then the multicast socket is turned on, would it be better?

**Comment by Logan007** [comment]
> There are too many options in n2n, and it seems too complicated, add a great cognitive burden to users. 

I fully agree. A lot of them are expert settings. If we had graphical cponfiguration, I would hide a lot of them on an extra page for advanced users.

> For this problem, if the `-E` option is not added, do not enable the multicast `socket eee->udp_multicast_sock`, and then turn off all multicast-related functions. If the `-E` option is added, and then the multicast socket is turned on, would it be better?

I would not want to merge TAP-related options (`-E`) concerning the virtual network with those configuring the underlying real network behavior.

> Would it be better if similar and unimportant functions are merged into the same option?

Generally spoken, yes.  However, I see potential for other options to be merged, namely `-b` into `-p` (requires some parsing). This is something I want to do before the release. Let me  just open an issue on cli options to discuss details over there.

**Comment by Logan007** [comment]
Please note that #772 merges `-b` into `-p`.
